app/domain/external_invoice/sac_membership/membership_manager.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+#  Copyright (c) 2024, Schweizer Alpen-Club. This file is part of
+#  hitobito_sac_cas and licensed under the Affero General Public License version 3
+#  or later. See the COPYING file at the top-level directory or at
+#  https://github.com/hitobito/hitobito_sac_cas
+
+class ExternalInvoice::SacMembership::MembershipManager
+  def initialize(person, group, year)
+    @person = person
+    @group = group
+    @year = year
+
+    context = Invoices::SacMemberships::Context.new(Time.zone.today)
+    @member ||= Invoices::SacMemberships::Member.new(person, context)
+  end
+
+  def update_membership_status
+    ActiveRecord::Base.transaction do
+      if stammsektion?
+        extend_membership_duration(@person, @year)
+      elsif neuanmeldung_stammsektion?
+        set_confirmed_at
+        update_role_to_stammsektion_mitglied(@person)
+        @person.household_people.each { |family_member| update_role_to_stammsektion_mitglied(family_member) } if family_main_person?
+      elsif neuanmeldung_zusatzsektion?
+        update_roles_to_zusatzsektion_mitglied(@person)
+        @person.household_people.each { |family_member| update_roles_to_zusatzsektion_mitglied(family_member) } if family_main_person?
+      end
+    end
+  end
+
+  private
+
+  def extend_membership_duration(person, year)
+    roles_for_update = []
+
+    roles_for_update << person.sac_membership.stammsektion_role
+    roles_for_update.concat(person.sac_membership.zusatzsektion_roles.reject(&:terminated?))
+    if family_main_person?
+      roles_for_update.concat(family_memberships.map(&:stammsektion_role))
+      roles_for_update.concat(family_memberships.flat_map(&:zusatzsektion_roles)
+                      .select { |zusatzsektion| zusatzsektion.beitragskategorie == "family" })
+    end
+
+    roles_for_update.each do |role|
+      role.update!(delete_on: [Date.new(year).end_of_year, role.delete_on].max)
+    end
+  end
+
+  def update_role_to_stammsektion_mitglied(person)
+    person.sac_membership.neuanmeldung_stammsektion_role.destroy
+    create_mitglied_role(person)
+  end
+
+  def update_roles_to_zusatzsektion_mitglied(person)
+    person.sac_membership.neuanmeldung_zusatzsektion_roles.each do |role|
+      role.destroy
+      create_mitglied_zusatzsektion_role(person)
+    end
+  end
+
+  def create_mitglied_role(person)
+    Group::SektionsMitglieder::Mitglied.create!(person: person, group: @group.layer_group.children.where(type: Group::SektionsMitglieder.sti_name).first, delete_on: Date.new(@year).end_of_year, created_at: Time.zone.now)
```

Comment:
`@group.layer_group.children.where(type: Group::SektionsMitglieder.sti_name).first` bitte auch noch in eine method extrahieren, wird bei `create_mitglied_zusatzsektion_role` wieder verwendet